PSReadLine/Prediction.Views.cs

```diff
@@ -0,0 +1,780 @@
+                                    int matchIndex = sugText.IndexOf(_inputText, comparison);
+                                    _listItems.Add(new SuggestionEntry(item.Name, item.Id, sugText, matchIndex));
+                                }
+                            }
+                        }
+                    }
+                    finally
+                    {
+                        _cacheList1.Clear();
+                        _cacheList2.Clear();
+                    }
+                }
+
+                if (_listItems?.Count > 0)
+                {
+                    _listItemHeight = Math.Min(_listItems.Count, ListMaxCount);
+                }
+                else
+                {
+                    Reset();
+                }
+            }
+
+            internal override void RenderSuggestion(List<StringBuilder> consoleBufferLines, ref int currentLogicalLine)
+            {
+                if (_updatePending)
+                {
+                    _updatePending = false;
+                }
+                else
+                {
+                    AggregateSuggestions();
+                }
+
+                if (_listItems == null)
+                {
+                    return;
+                }
+
+                for (int i = 0; i < _listItemHeight; i++)
+                {
+                    currentLogicalLine += 1;
+                    if (currentLogicalLine == consoleBufferLines.Count)
+                    {
+                        consoleBufferLines.Add(new StringBuilder(COMMON_WIDEST_CONSOLE_WIDTH));
+                    }
+
+                    bool itemSelected = i == _selectedIndex;
+                    StringBuilder currentLineBuffer = consoleBufferLines[currentLogicalLine];
+
+                    string selectionColor = itemSelected ? _singleton._options._listPredictionSelectedColor : null;
+                    currentLineBuffer.Append(
+                        _listItems[i].GetListItemText(
+                            _listItemWidth,
+                            _inputText,
+                            selectionColor));
+
+                    if (itemSelected)
+                    {
+                        currentLineBuffer.Append(VTColorUtils.AnsiReset);
+                    }
+                }
+            }
+
+            internal override void OnSuggestionAccepted()
+            {
+                if (!UsePlugin)
+                {
+                    return;
+                }
+
+                if (_listItems != null && _selectedIndex != -1)
+                {
+                    var item = _listItems[_selectedIndex];
+                    if (item.PredictorId != Guid.Empty)
+                    {
+                        _singleton._mockableMethods.OnSuggestionAccepted(item.PredictorId, item.SuggestionText);
+                    }
+                }
+            }
+
+            internal override void Clear(bool cursorAtEol)
```

Comment:
Instead of passing in `cursorAtEol`, would it be better to have a helper calculate it?

Reply:
This parameter exists for perf reason. To clear the view, we need to the `CursorTop` if the cursor is at the end of the input text. But in some common scenarios, the cursor is already at the end and there is no need to calculate that again.